### DISCUSSION TITLE
Is there anyway to use server actions to provide language client connections?

### DISCUSSION THREAD

**Discussion by alissa-tung** [TypeFox/monaco-languageclient#606] [Q&A]
In recent versions of React allowed some functions to run on server side runtime (for example, node), and pass the result of calling such functions to client side runtime (browser)

A common language server program can be called by using NodeJS `child_process` spawn, where the return value of such `spawn` have two properties `stdin` and `stdout`.

stdin and stdout can be turns in to `StreamMessageReader` and `StreamMessageWriter` from `vscode-jsonrpc/node`, the pair of them is `MessageTransport`

the key of Message reader and writer is the `read` and `write` function, that is, consider the following case

- listen: can be called only once, started listen to following messages, takes a callback as argument: listen : (JSON -> void) -> Disposable
- write: JSON -> Promise<void>

if we had already created StreamMessageReader and StreamMessageWriter on server side, export the write and read function as async server actions, then in the client side we declare two new class, use the exported server actions to define the methods

```ts
class ServerStreamMessageReader
  extends rpc.AbstractMessageReader
  implements rpc.MessageReader
{
  listen(callback: rpc.DataCallback): rpc.Disposable {
```

and

```ts
class ServerStreamMessageWriter
  extends rpc.AbstractMessageWriter
  implements rpc.MessageWriter
{
  write(msg: rpc.Message): Promise<void> {
```

but directly wrap the server actions is not okay, since server actions can not have a client function as argument.

So can we use a API which is not like `listen` (which takes a callback) , or can we have some methods to fix the server actions?

**Comment by kaisalmen** [comment]
Hi @alissa-tung I have no experience with React Server Functions, so I can't give any recommendation there.

For the distributed case (=client and server in a different environment) we use `vscode-ws-jsonrpc`. Is it not possible to open a websocket in your environment?

**Comment by alissa-tung** [comment]
> Hi @alissa-tung I have no experience with React Server Functions, so I can't give any recommendation there.
> 
> For the distributed case (=client and server in a different environment) we use `vscode-ws-jsonrpc`. Is it not possible to open a websocket in your environment?

Thank you, I will read something about those libraries. My intuitions are

- make deploy easier, since my environment is not easy to expose web socket connections to end user
- I would like to move the Monaco language client relevant code entirely to the server side, since server actions are fast (less network redirection), and I would hope end user to run as less JavaScript as possible, I hope my software can work in very retro devices
- not expose lsp connection to end user may safer: if user is allowed to send any message to the language server, there may occur some unexpected event to there code and my server

**Comment by kaisalmen** [comment]
> and I would hope end user to run as less JavaScript as possible, I hope my software can work in very retro devices

Generally, `monaco-editor` requires client-side JavaScript and lots of it, so removing the need for it in the language client is not really changing anything with regard to retro devices.

> not expose lsp connection to end user may safer: if user is allowed to send any message to the language server, there may occur some unexpected event to there code and my server

The language server is something isolated and not so much a "server" (the naming is unfortunately misleading), but more an input/output processor. you should always see this as a seperate part from your regular server. Don't know if you are aware that a language server can run in a web worker (client side), but this is only helpful if your language server can run in a browser environment.

**Comment by alissa-tung** [comment]
> > and I would hope end user to run as less JavaScript as possible, I hope my software can work in very retro devices
> 
> Generally, `monaco-editor` requires client-side JavaScript and lots of it, so removing the need for it in the language client is not really changing anything with regard to retro devices.
> 
> > not expose lsp connection to end user may safer: if user is allowed to send any message to the language server, there may occur some unexpected event to there code and my server
> 
> The language server is something isolated and not so much a "server" (the naming is unfortunately misleading), but more an input/output processor. you should always see this as a seperate part from your regular server. Don't know if you are aware that a language server can run in a web worker (client side), but this is only helpful if your language server can run in a browser environment.

Thank you for responding. We are working around a special language server, where there are many aggregation operations using the code of language client, so the performance is not good like other (CPP, Java, Python), as they only needs to call code action, completion and so on. But in our case there are many aggregations. It is not convenient to move them to lsp custom request/notification for now.

As for the safety part, when the language server receives invalid request it will crash directly, which can lose state on the current lsp server process, although make every connections is auth can solve this problem, but we are trying to keep a public playground for unlogined users. And existed custom notification/requests can execute any code. (anyway, we had built a bwrap wrapper for now)

**Comment by alissa-tung** [comment]
I realized this is not an issue, maybe we can move to discussion page. So close it for now
